homeassistant/components/bluetooth/__init__.py

```diff
@@ -227,14 +273,22 @@ def _device_detected(
     ) -> None:
         """Handle a detected device."""
```

Comment:
For this method, why do we keep matching domains if there is a callback? I think that we should be able to assume that if a callback exists, that's where the data goes. We can index the address listening callbacks by address to quickly look it up, forward data. 

So then it would be:

- Packet comes in
- Callback for address? Check if it matches and forward it. Callback can still filter on extra data if it knows it doesn't care about all packets from that device.
- Seen this device before? Yes, drop packet
- Mark as seen
- Match packet on integrations

Reply:
For registered callbacks: not all integrations are going to want callbacks by address.  Some will want device with a specific UUID for beacons.

For domain flows:  We only match on the same address maximum 2x.  Once for the no `manufacturer_data` case (some integrations are fine without it since they only care about the service uuids), and again for the has `manufacturer_data` case (some won't get anything without it)